Add post build-a-crud-app-with-jsf (#33)
oktadeveloper/okta-blog: _source/_posts/2019-09-27-build-a-crud-app-with-jsf.md

---
layout: blog_post

> title: "Build a Simple Application with JavaServer Faces"

Reviewer: Should CRUD be in the title? It's in the URL and description. It looks like the keywords are `java jsf` and neither is in the title. `Build a CRUD App with Java and JSF` is a 57 on https://brentisarulebreaker.dev/

```xml
<h:commandButton action="#{bookBean.fetchBooks}" value="View books present"/>
```

This executes the method `fetchBooks()` in the class. In the `src/main/java/com/okta/developer/presentation/BookBean.java` file, we see `fetchBooks()` delegates to a method in `bookService`, which stores the results into the `booksAvailable` field, and then returns the string "success".

```java
public String fetchBooks() {
    booksAvailable=bookService.getAllBooks();
    return "success";
}
```

## Read the books in your JSF Application

In the `getAllBooks()` method, the `com.okta.developer.application.BookService` class queries the database fetching all `Book` instances with no filters:

```java
public List<Book> getAllBooks() {
    CriteriaQuery<Book> cq = entityManager.getCriteriaBuilder().createQuery(Book.class);
    cq.select(cq.from(Book.class));
    return entityManager.createQuery(cq).getResultList();
}
```

Cool. But how does the page actually display the book information? In the `result.xhtml` file, find the `ui:repeat` tag:

```xml
<ui:repeat value="#{bookBean.booksAvailable}" var="book">
    #{book.bookTitle} <br/>
</ui:repeat>
```

> The `ui:repeat` tag iterates over its `value`, in this case `#{bookBean.booksAvailable` the field we just assigned from the `fetchBooks()` method. Each element of the collection can be referenced by the name in the `var` attribute of the tag (in this case `book`).

Reviewer: ```suggestion
The `<ui:repeat>` tag iterates over each `value`, in this case `#{bookBean.booksAvailable}` is the field we just assigned from the `fetchBooks()` method. Each element of the collection can be referenced by the name in the `var` attribute of the tag (in this case `book`).
```

> Whatever is inside the `ui:repeat` tag will be repeated for each of the elements in the collection. Here, it simply outputs the book title by using the interpolation notation `#{book.bookTitle}` followed by a line break tag (`<br/>`).

Reviewer: ```suggestion
Whatever is inside the `<ui:repeat>` tag will be repeated for each of the elements in the collection. Here, it simply outputs the book title by using the interpolation notation `#{book.bookTitle}` followed by a line break tag (`<br/>`).
```

    CriteriaQuery<Book> cq = cb.createQuery(Book.class);
    Root<Book> book = cq.from(Book.class);
    cq.select(book);
    cq.where(cb.equal(book.get("bookId"), bookId));
    return entityManager.createQuery(cq).getSingleResult();
}
```

And add the following logic to `BookBean` to load the book before page render:

```java
import javax.annotation.PostConstruct;
import javax.faces.context.FacesContext;

...

private Integer bookId;
private Book book;

@PostConstruct
public void postConstruct() {
    String bookIdParam = FacesContext.getCurrentInstance()
        .getExternalContext().getRequestParameterMap().get("bookId");
    if (bookIdParam != null) {
        bookId = Integer.parseInt(bookIdParam);
        book = bookService.getBook(bookId);
        bookTitle = book.getBookTitle();
    }
}
```

Now let's create a method that will persist the changes to the database, with the following method in the `BookService` class:

```java
public void update(Book book) {
    entityManager.merge(book);
}
```

Also, add the `update` method to `BookBean` class:

```java
public String update() {
    book.setBookTitle(bookTitle);
    bookService.update(book);
    return "success";
}
```

> And to properly redirect the user to the list page, we add the following to the file `src/main/webapp/faces-config.xml`:

Reviewer: ```suggestion
And to properly redirect the user to the list page, add the following navigation rule to the file `src/main/webapp/faces-config.xml`:
```

```xml
<navigation-rule>
  <from-view-id>/edit.xhtml</from-view-id>
  <navigation-case>
    <from-outcome>success</from-outcome>
    <to-view-id>/result.xhtml</to-view-id>
  </navigation-case>
</navigation-rule>
```

Now that we are done updating the book, let's move on to the Delete part.

## Delete A Record

> The previous section covers hard part - loading a book inside the bean.Adding the delete button will be easier.

Reviewer: ```suggestion
The previous section covers the hard part - loading a book inside the bean. Adding the delete button will be easier.
```

    </h:link>

    <!-- Delete link: -->
    <h:outputText value=" ("/>
    <h:link value="Delete" outcome="delete">
        <f:param name="bookId" value="#{book.bookId}"/>
    </h:link>
    <h:outputText value=")"/>
    <br/>
</ui:repeat>
```

Now,create the delete confirmation page at `src/main/webapp/delete.xhtml`:

```xml
<?xml version="1.0" encoding="UTF-8"?>

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
        "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"
      xmlns:f="http://java.sun.com/jsf/core"
      xmlns:h="http://java.sun.com/jsf/html">

<h:body bgcolor="white">
    <f:view>
        <h1>Delete book?</h1>
        <h:form>
            <h:panelGrid columns="2">
                <h:outputText value='Book title'/>
                <h:inputText value='#{bookBean.bookTitle}' readonly="true"/>
                <h:outputText value='Delete'/>
                <h:commandButton action="#{bookBean.delete}" value="Confirm Delete"/>
            </h:panelGrid>
            <input type="hidden" name="bookId" value='#{param.bookId}'/>
        </h:form>
    </f:view>
</h:body>
</html>
```

> And add the proper delete handler in our `BookBean` class:

Reviewer: ```suggestion
And add the proper delete handler in the `BookBean` class:
```

import com.okta.developer.entities.Book;

import javax.annotation.PostConstruct;
import javax.inject.Inject;
import javax.inject.Named;
import java.util.List;

@Named
public class BookList {

    @Inject
    private BookService bookService;
    private List<Book> booksAvailable;

    @PostConstruct
    public void postConstruct() {
        booksAvailable = bookService.getAllBooks();
    }

    public List<Book> getBooksAvailable() {
        return booksAvailable;
    }
}
```

Remove the following blocks of code related to the `booksAvailable` field from `BookBean` class: 

```java
private List<Book> booksAvailable;

public List<Book> getBooksAvailable() {
    return booksAvailable;
}

public void setBooksAvailable(List<Book> booksAvailable) {
    this.booksAvailable = booksAvailable;
}

public String fetchBooks() {
    booksAvailable=bookService.getAllBooks();
    return "success";
}
```

Let's also change our landing page. Instead of adding a Book through the form, let's present the book list. For that, edit `index.jsp` to change the redirect to `result.jsf`:

```jsp
<%@ page session="false" %>
<%
    response.sendRedirect("result.jsf");
%>
```

> Here is what my files now look like. Feel free to adapt yours as you browse the PrimeFaces component library.

Reviewer: It seems like there should be a link around `PrimeFaces component library`

        </p:link>
        <h:outputText value=")"/>
    </p:dataList>
</h:body>
</html>
```

## Run the Application with PrimeFaces Enabled

Restart the app with `mvn package tomee:run`. The app will look a bit better now! Check out the book list:

{% img blog/build-a-crud-app-with-jsf/prime-faces-list.png alt:"Book list with PrimeFaces" width:"600" %}{: .center-image }

## Secure Your Application With Okta

At the moment, anyone can access our awesome Book application and change the database.
To prevent that, let's add a security layer to our application with the Spring Security library and authenticate users through Okta.

First, [register for a forever-free developer account today!](https://developer.okta.com/signup/) When you're finished, complete the steps below to create an OIDC app.

1. Log in to your developer account at [developer.okta.com](https://developer.okta.com/).
2. Navigate to **Applications** and click on **Add Application**.
3. Select **Web** and click **Next**.
4. Give the application a name (.e.g., `Java JSF Secure CRUD`) 
5. Add the following as Login redirect URI:
    - `http://localhost:8080/login/oauth2/code/okta`

> 5. Click **Done**

Reviewer: I'd remove the periods from the first three bullets since the last three don't have them.
